File: codigo_ia/src/lag_transformer.py
from __future__ import annotations

from typing import List, Optional, Sequence

import pandas as pd
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.utils.validation import check_is_fitted


class LagByGroupDateTransformer(BaseEstimator, TransformerMixin):
    """
    Crea una columna de lag del agregado de `lag_column` por
    (nivel_agregacion, date_col), desplazando n días hacia atrás y uniendo
    al dataset original.

    - fit: calcula y guarda la tabla de lookup de lags (NO toca X).
    - transform: hace merge de X con la tabla de lags ya calculada
      (sin recomputar).
    """

    def __init__(
        self,
        nivel_agregacion: Sequence[str],
        n: int = 1,
        agg_func: str = "sum",
        date_col: str = "Date",
        lag_column: str = "Units Sold",
        ref_date: Optional[pd.Timestamp | str] = None,
        keep_original_date: bool = True,
    ) -> None:
        self.n = int(n)
        self.nivel_agregacion = list(nivel_agregacion)
        self.agg_func = agg_func
        self.date_col = date_col
        self.lag_column = lag_column
        self.ref_date = pd.to_datetime(ref_date).normalize()
        self.keep_original_date = keep_original_date

    def _validate_input(self, X: pd.DataFrame) -> pd.DataFrame:
        X = X.copy()
        if self.date_col not in X.columns:
            raise ValueError(
                (f"'{self.date_col}' no está en columnas del " f"DataFrame.")
            )
        for col in self.nivel_agregacion + [self.lag_column]:
            if col not in X.columns:
                raise ValueError(f"'{col}' no está en columnas del DataFrame.")
        X[self.date_col] = pd.to_datetime(X[self.date_col])
        return X

    def _build_lookup(self, X: pd.DataFrame) -> pd.DataFrame:
        # Agrega por grupo + fecha
        # Preparamos el nombre de la columna agregada antes de agrupar
        # (evita problemas de tipo)
        label = "_".join(self.nivel_agregacion)
        lag_feat = f"lag_{self.lag_column}_{label}_{self.agg_func}_{self.n}"

        grp = (
            X.groupby(self.nivel_agregacion + [self.date_col], dropna=False)[
                self.lag_column
            ]
            .agg(self.agg_func)
            .reset_index(name=lag_feat)
        )

        # Desplaza la fecha hacia ATRÁS n días para construir la clave join
        grp["date_lag"] = grp[self.date_col] - pd.Timedelta(days=self.n)
        if self._effective_ref_date is not None:
            grp = grp.loc[grp[self.date_col] <= self._effective_ref_date]
        # Nos quedamos con las columnas necesarias para el merge en transform
        lookup_cols = self.nivel_agregacion + ["date_lag", lag_feat]
        return (
            grp[lookup_cols]
            .sort_values(self.nivel_agregacion + ["date_lag"])
            .reset_index(drop=True)
        )

    def fit(
        self, X: pd.DataFrame, y: Optional[pd.Series] = None
    ) -> "LagByGroupDateTransformer":
        X = self._validate_input(X)

        # ref_date efectiva: si no se pasó, tomamos hoy
        self._effective_ref_date = (
            self.ref_date
            if self.ref_date is not None
            else pd.Timestamp.today().normalize()
        )

        # Construye y guarda la tabla lookup
        self._lag_feature_name_ = (
            f"lag_{self.lag_column}_"
            f"{'_'.join(self.nivel_agregacion)}_{self.agg_func}_{self.n}"
        )
        self._lookup_df_ = self._build_lookup(X)

        # Guarda un set de columnas para chequeos
        self._input_columns_ = list(X.columns)
        return self

    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
        check_is_fitted(self, attributes=["_lookup_df_", "_lag_feature_name_"])
        X = self._validate_input(X)

        # Clonamos para no tocar el original
        out = X.copy()

        # Si se quiere, recortamos los datos de entrada a fechas que no miren
        # al futuro respecto de la ref_date efectiva (evita fugas con fechas
        # futuras en X).

        # Preparamos claves de merge: izquierda usa
        # (nivel_agregacion + date_col), derecha usa
        # (nivel_agregacion + "date_lag") para traer el valor de hace n días.
        right_on = self.nivel_agregacion + ["date_lag"]
        left_on = self.nivel_agregacion + [self.date_col]

        out = out.merge(
            self._lookup_df_,
            how="left",
            left_on=left_on,
            right_on=right_on,
            sort=False,
        )

        # Limpieza de columnas auxiliares
        if not self.keep_original_date:
            out = out.drop(columns=[self.date_col], errors="ignore")
        out = out.drop(columns=["date_lag"], errors="ignore")

        return out

    def fit_transform(
        self,
        X: pd.DataFrame,
        y: Optional[pd.Series] = None,
        **fit_params: object,
    ) -> pd.DataFrame:
        return self.fit(X, y).transform(X)

    # Conveniencia
    def get_feature_names_out(
        self, input_features: Optional[List[str]] = None
    ) -> List[str]:
        check_is_fitted(self, attributes=["_lag_feature_name_"])
        return [self._lag_feature_name_]

File: codigo_ia/src/lag_transformer.R
#' LagByGroupDateTransformerR: transformador de lags por grupo y fecha
#'
#' Crea una característica lag agregada (sum/mean/...) por (`nivel_agregacion`, `date_col`),
#' desplazada `n` días hacia atrás, y la une al dataset original.
#'
#' - fit(): calcula y guarda la tabla de lookup (no modifica X)
#' - transform(): hace join de X con la tabla de lags
#'
suppressPackageStartupMessages({
  requireNamespace("R6", quietly = TRUE)
  requireNamespace("dplyr", quietly = TRUE)
  requireNamespace("lubridate", quietly = TRUE)
})

LagByGroupDateTransformerR <- R6::R6Class( # nolint: object_name_linter
  classname = "LagByGroupDateTransformerR",
  public = list(
    nivel_agregacion = NULL,
    n = NULL,
    agg_func = NULL,
    date_col = NULL,
    lag_column = NULL,
    ref_date = NULL,
    keep_original_date = NULL,
    initialize = function(nivel_agregacion,
                          n = 1,
                          agg_func = "sum",
                          date_col = "Date",
                          lag_column = "Units Sold",
                          ref_date = NULL,
                          keep_original_date = TRUE) {
      stopifnot(is.character(nivel_agregacion), length(nivel_agregacion) >= 1)
      self$nivel_agregacion <- nivel_agregacion
      self$n <- as.integer(n)
      self$agg_func <- match.arg(agg_func, c("sum", "mean", "median", "max", "min"))
      self$date_col <- as.character(date_col)
      self$lag_column <- as.character(lag_column)
      self$ref_date <- if (!is.null(ref_date)) as.Date(ref_date) else NULL
      self$keep_original_date <- isTRUE(keep_original_date)
    },
    fit = function(x) {
      x <- private$validate_input(x)
      private$effective_ref_date <- if (!is.null(self$ref_date)) self$ref_date else Sys.Date()

      label <- paste(self$nivel_agregacion, collapse = "_")
      private$lag_feature_name <- sprintf(
        "lag_%s_%s_%s_%d",
        self$lag_column, label, self$agg_func, self$n
      )

      private$lookup_df <- private$build_lookup(x)
      private$input_columns <- colnames(x)
      invisible(self)
    },
    transform = function(x) {
      if (is.null(private$lookup_df) || is.null(private$lag_feature_name)) {
        stop("El transformador no está ajustado. Llama a fit() primero.")
      }
      x <- private$validate_input(x)

      # Join: izquierda (X) con derecha (lookup) mapeando date_col <- date_lag
      by_cols <- c(self$nivel_agregacion, self$date_col)
      right <- private$lookup_df
      # renombrar date_lag a date_col para que el join sea directo
      right2 <- dplyr::rename(right, !!self$date_col := .data$date_lag)

      out <- dplyr::left_join(x, right2, by = by_cols)

      if (!self$keep_original_date) {
        out <- dplyr::select(out, -dplyr::all_of(self$date_col))
      }
      out
    },
    fit_transform = function(x) {
      self$fit(x)
      self$transform(x)
    },
    get_feature_names_out = function() {
      if (is.null(private$lag_feature_name)) {
        stop("El transformador no está ajustado. Llama a fit() primero.")
      }
      private$lag_feature_name
    }
  ),
  private = list(
    input_columns = NULL,
    lookup_df = NULL,
    lag_feature_name = NULL,
    effective_ref_date = NULL,
    validate_input = function(x) {
      stopifnot(is.data.frame(x))
      cols_needed <- c(self$nivel_agregacion, self$date_col, self$lag_column)
      falta <- setdiff(cols_needed, colnames(x))
      if (length(falta) > 0) {
        stop(sprintf("Faltan columnas en X: %s", paste(falta, collapse = ", ")))
      }
      x[[self$date_col]] <- as.Date(x[[self$date_col]])
      x
    },
    build_lookup = function(x) {
      label <- paste(self$nivel_agregacion, collapse = "_")
      lag_feat <- private$lag_feature_name

      agg_fun <- switch(self$agg_func,
        sum = sum,
        mean = mean,
        median = median,
        max = max,
        min = min
      )

      grp <- x |>
        dplyr::group_by(dplyr::across(dplyr::all_of(c(self$nivel_agregacion, self$date_col)))) |>
        dplyr::summarise(
          .agg = agg_fun(.data[[self$lag_column]], na.rm = TRUE),
          .groups = "drop"
        ) |>
        dplyr::rename(!!lag_feat := .data$.agg)

      grp <- grp |>
        dplyr::mutate(date_lag = .data[[self$date_col]] - lubridate::days(self$n))

      if (!is.null(private$effective_ref_date)) {
        grp <- dplyr::filter(grp, .data[[self$date_col]] <= private$effective_ref_date)
      }

      lookup_cols <- c(self$nivel_agregacion, "date_lag", lag_feat)
      grp |>
        dplyr::select(dplyr::all_of(lookup_cols)) |>
        dplyr::arrange(dplyr::across(dplyr::all_of(c(self$nivel_agregacion, "date_lag")))) |>
        as.data.frame()
    }
  )
)
